wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,595 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Region meta data
+
+use std::{
+    cmp,
+    collections::{BTreeMap, HashMap},
+    sync::atomic::{AtomicI64, Ordering},
+};
+
+use common_types::{table::TableId, SequenceNumber};
+use common_util::define_result;
+use message_queue::Offset;
+use snafu::{ensure, Backtrace, OptionExt, Snafu};
+use tokio::sync::{Mutex, RwLock};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to get meta data of table:{}, msg:{}\nBacktrace:{}",
+        table_id,
+        msg,
+        backtrace
+    ))]
+    GetMetaData {
+        table_id: TableId,
+        msg: String,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Failed to update meta data after write of table:{}, msg:{}\nBacktrace:{}",
+        table_id,
+        msg,
+        backtrace
+    ))]
+    UpdateAfterWrite {
+        table_id: TableId,
+        msg: String,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Failed to mark deleted for table:{}, msg:{}\nBacktrace:{}",
+        table_id,
+        msg,
+        backtrace
+    ))]
+    MarkDeleted {
+        table_id: TableId,
+        msg: String,
+        backtrace: Backtrace,
+    },
+}
+define_result!(Error);
+
+// TODO: will be made use of later.
+#[allow(unused)]
+#[derive(Default, Debug)]
+pub struct RegionMeta {
+    inner: RwLock<RegionMetaInner>,
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+impl RegionMeta {
+    // TODO: Need to implement the init method using the [RegionMetaSnapshot] which
+    // will be persisted.
+
+    pub async fn prepare_for_table_write(&self, table_id: TableId) -> Result<SequenceNumber> {
+        {
+            let inner = self.inner.read().await;
+            if let Some(table_meta) = inner.table_metas.get(&table_id) {
+                return Ok(table_meta.get_meta_data().await?.next_sequence_num);
+            }
+        }
+
+        // Double check is not needed, due to just one task will write the specific
+        // table.
+        let mut inner = self.inner.write().await;
+        debug_assert!(inner
```

Comment:
I propose to keep the logic and just add an assert now.
I want to let making it adapt to other thread model be an enhancement which will be implemented later. Because for reaching it, the key task should be modifying the lock granularity but not the logic here. As I see, it may be not so suitable to do in one pr.